Use an isLoading flag to prevent fetchHomeData from being dispa…

bobsilverberg opened the pull request:
Fixes #8572

kumar303 commented:
Disabling strict mode made this behave as you expected, right? The logic looks correct to me. Your observations correlate with what I'd think would happen in strict mode.

bobsilverberg commented:
> Disabling strict mode made this behave as you expected, right? 

Yes, that resolved my issue, thanks! This is now ready for a full review.

kumar303 commented on src/amo/sagas/home.js:
```diff
@@ -136,6 +142,7 @@ export function* fetchHomeData({
       });
     } catch (error) {
       log.warn(`Home add-ons failed to load: ${error}`);
+      yield put(abortFetchHomeData());
```
You don't need this because the error will get caught in an outer block which already aborts the fetch. It would be good to keep the extra test for it, though.

kumar303 commented on src/amo/sagas/home.js:
```diff
@@ -46,6 +50,7 @@ export function* fetchHomeData({
       heroShelves = yield call(getHeroShelves, { api: state.api });
     } catch (error) {
       log.warn(`Home hero shelves failed to load: ${error}`);
+      yield put(abortFetchHomeData());
```
Actually, you don't need this one either for the same reason.

kumar303 commented on src/amo/sagas/home.js:
```diff
@@ -68,6 +73,7 @@ export function* fetchHomeData({
           // The collection was not found or is marked private.
           collections.push(null);
         } else {
+          yield put(abortFetchHomeData());
```
nor this one

kumar303 commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -421,6 +444,41 @@ describe(__filename, () => {
+      const firstCollection = createFakeCollectionAddonsListResponse();
+      mockCollectionsApi
+        .expects('getCollectionAddons')
+        .withArgs({
```
The arguments are already covered by other tests so you can remove this line. Also, `mockCollectionsApi.verify()` is never called (which is fine) so this doesn't have an affect.

kumar303 commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -421,6 +444,41 @@ describe(__filename, () => {
+
+      mockSearchApi
+        .expects('search')
+        .exactly(5)
```
You can remove this since it won't have an affect (`mockSearchApi.verify()` is not called). It's not important for this test, anyway, which is about handling errors.

kumar303 commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -421,6 +444,41 @@ describe(__filename, () => {
+          slug,
+          userId,
+        })
+        .returns(Promise.resolve(firstCollection));
```
I think you can do `.resolves()`

kumar303 commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -421,6 +444,41 @@ describe(__filename, () => {
+      mockSearchApi
+        .expects('search')
+        .exactly(5)
+        .returns(Promise.reject(error));
```
I think you can do `.rejects()`

kumar303 commented on tests/unit/amo/sagas/test_home.js:
```diff
@@ -432,5 +490,18 @@ describe(__filename, () => {
+    it('aborts fetching for a failed hero fetch', async () => {
+      const error = createApiError({ response: { status: 500 } });
+
+      mockHeroApi.expects('getHeroShelves').returns(Promise.reject(error));
```
This could use `.rejects()`.

bobsilverberg commented:
Thanks @kumar303. I have addressed all of your comments. I noticed that some of the existing tests could also be cleaned up a bit. Both by replacing instances of  `returns(Promise.reject())` with `rejects()`, and same with `resolves`; and also by removing some of the expectations on the mockApis where we are not later verifying the API call.

I didn't make those changes because I did want to make the patch unnecessarily large, but I'm thinking that maybe I should make those changes before landing this, after you've had a chance to do another review. WDYT?

kumar303 commented:
> I didn't make those changes because I did want to make the patch unnecessarily large

I agree. There is so much cleanup needed in addons-frontend but it's impractical to squeeze into every PR.

bobsilverberg commented on src/amo/reducers/home.js:
```diff
@@ -251,9 +265,16 @@ const reducer = (
     case FETCH_HOME_DATA:
       return {
         ...state,
+        isLoading: true,
```
@kumar303 There is this test: https://github.com/mozilla/addons-frontend/blob/d9d68a0b19f376340b60be940039debc3542506f/tests/unit/amo/reducers/test_home.js#L214-L227

Rather than adding a separate test just to check the value of `isLoading`, I updated an existing test to verify that both `isLoading` and `resultsLoaded` are correct after a fetch. This seems okay to me because it represents the expected state for a a single action.

Do you feel it's important to have separate tests for both the `resultsLoaded` and the `isLoading` flags, or did you perhaps miss noticing this change in your review?

bobsilverberg replied:
@kumar303 I'm just waiting on an answer to this question before merging this patch. Thanks!

kumar303 replied:
Ship it!
